### PR TITLE
configurable log process thread priority

### PR DESCRIPTION
**Pull request by justinyhuang**
added Zephyr config items to support setting the priority of the log thread at build time.

Signed-off-by: Justin Huang <justinyhuang@fb.com>

### CODE REVIEW

**Comment by justinyhuang** [comment]
@cfriedt Would you mind sharing your thoughts on this draft PR?
I will publish a formal PR once it looks good to you.

**Review by cfriedt — DISMISSED** [review]
LGTM

**Comment by cfriedt** [comment]
cc @nordic-krch

**Comment by justinyhuang** [comment]
update the commit to address some format issues

File: subsys/logging/Kconfig.processing [hunk 1]
```diff
@@ -90,6 +90,17 @@ config LOG_PROCESS_THREAD_STACK_SIZE
 	help
 	  Set the internal stack size for log processing thread.
 
+config LOG_PROCESS_THREAD_SPECIFY_THREAD_PRIORITY
+	bool "Enable specifying the priority of the internal log processing thread"
```
Comment (nordic-krch) [inline subsys/logging/Kconfig.processing:94]: There was an action to remove prompts starting with "Enable ..." #43586. Maybe something like "Set custom thread priority". `internal` might not be needed as this is already dependent on option enabling internal thread.

File: subsys/logging/Kconfig.processing [hunk 2]
```diff
@@ -90,6 +90,17 @@ config LOG_PROCESS_THREAD_STACK_SIZE
...
+	default n
```
Comment (nordic-krch) [inline subsys/logging/Kconfig.processing:95]: it's redunant and can be removed.

File: subsys/logging/Kconfig.processing [hunk 3]
```diff
@@ -90,6 +90,17 @@ config LOG_PROCESS_THREAD_STACK_SIZE
...
+	help
+	  When enabled the priority of the log processing thread can be set at build time.
+	  The default is to set the priority to K_LOWEST_APPLICATION_THREAD_PRIO
```
Comment (nordic-krch) [inline subsys/logging/Kconfig.processing:98]: "The default" is not clear, maybe better would be "If not enabled priority is set to ..".

File: subsys/logging/log_core.c [hunk 4]
```diff
@@ -42,6 +42,13 @@ LOG_MODULE_REGISTER(log);
 #define CONFIG_LOG_BUFFER_SIZE 4
 #endif
 
+
+#if CONFIG_LOG_PROCESS_THREAD_SPECIFY_THREAD_PRIORITY
```
Comment (pdgendt) [inline subsys/logging/log_core.c:46]:
```suggestion
#ifdef CONFIG_LOG_PROCESS_THREAD_PRIORITY
```

Comment (cfriedt) [reply subsys/logging/log_core.c:46]: This suggestion works for me as well.

**Comment by cfriedt** [comment]
@justinyhuang - be sure to run `./scripts/checkpatch.pl -g HEAD` to catch any formatting issues (e.g. long line in commit message).

**Comment by justinyhuang** [comment]
I think I have updated the PR based on the feedback, but please let me know if I missed something.
Thanks!

**Comment by justinyhuang** [comment]
So I tried simplifing things by:

    config LOG_PROCESS_THREAD_PRIORITY
          int "Priority of the log processing thread"
          help
            The priority of the log processing thread.
            When not set the prority is set to K_LOWEST_APPLICATION_THREAD_PRIO.

and 

    #if IS_EMPTY(CONFIG_LOG_PROCESS_THREAD_PRIORITY)
    #define LOG_PROCESS_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
    #else
    #define LOG_PROCESS_THREAD_PRIORITY CONFIG_LOG_PROCESS_THREAD_PRIORITY
    #endif

but `samples/subsys/logging/logger` builds with error: 

    /var/folders/cb/tlt_12350v79w73_dmfks6v00000gn/T//ccO2wkIT.s: Assembler messages:
    /var/folders/cb/tlt_12350v79w73_dmfks6v00000gn/T//ccO2wkIT.s:1255: Error: missing expression

It looks like in this case `CONFIG_LOG_PROCESS_THREAD_PRIORITY` has to have a `default` value as `int` to make it build happily, which doesn't make sense here: the intention of having one config item, if I understand correctly here, is to use the value of the config item as the thread priority if set, and use the value of `K_LOWEST_APPLICATION_THREAD_PRIO` if it is not set. 

Thus I'd suggest to back the previous change, or please kindly suggest how to address the build failure above.

What do you think?

**Comment by cfriedt** [comment]
> So I tried simplifing things by:
> 
> ```
> config LOG_PROCESS_THREAD_PRIORITY
>       int "Priority of the log processing thread"
>       help
>         The priority of the log processing thread.
>         When not set the prority is set to K_LOWEST_APPLICATION_THREAD_PRIO.
> ```
> 
> and
> 
> ```
> #if IS_EMPTY(CONFIG_LOG_PROCESS_THREAD_PRIORITY)
> #define LOG_PROCESS_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
> #else
> #define LOG_PROCESS_THREAD_PRIORITY CONFIG_LOG_PROCESS_THREAD_PRIORITY
> #endif
> ```

I think that specific location you would need to use `#ifdef`, so

```
#ifdef CONFIG_LOG_PROCESS_THREAD_PRIORITY
#define LOG_PROCESS_THREAD_PRIORITY CONFIG_LOG_PROCESS_THREAD_PRIORITY
#else
#define LOG_PROCESS_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
#endif
```

**Comment by pdgendt** [comment]
> I think that specific location you would need to use `#ifdef`, so
> 
> ```
> #ifdef CONFIG_LOG_PROCESS_THREAD_PRIORITY
> #define LOG_PROCESS_THREAD_PRIORITY CONFIG_LOG_PROCESS_THREAD_PRIORITY
> #else
> #define LOG_PROCESS_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
> #endif
> ```

An `int` `Kconfig` option without a `default` value will result in the mentioned build error:
```
/tmp/ccY6ytUi.s: Assembler messages:
/tmp/ccY6ytUi.s:512: Error: missing expression
```

A boolean option to depend on, might still be required?

```
config LOG_PROCESS_THREAD_CUSTOM_PRIORITY
	bool "Custom log thread priority"
	help
	  Set a custom value for the log processing thread.

config LOG_PROCESS_THREAD_PRIORITY
	int "Priority of the log processing thread"
	default 0
	depends on LOG_PROCESS_THREAD_CUSTOM_PRIORITY
	help
	  The priority of the log processing thread.
	  When not set the prority is set to K_LOWEST_APPLICATION_THREAD_PRIO.
```

**Comment by justinyhuang** [comment]
> > I think that specific location you would need to use `#ifdef`, so
> > ```
> > #ifdef CONFIG_LOG_PROCESS_THREAD_PRIORITY
> > #define LOG_PROCESS_THREAD_PRIORITY CONFIG_LOG_PROCESS_THREAD_PRIORITY
> > #else
> > #define LOG_PROCESS_THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
> > #endif
> > ```
> 
> An `int` `Kconfig` option without a `default` value will result in the mentioned build error:
> 
> ```
> /tmp/ccY6ytUi.s: Assembler messages:
> /tmp/ccY6ytUi.s:512: Error: missing expression
> ```
> 
> A boolean option to depend on, might still be required?
> 
> ```
> config LOG_PROCESS_THREAD_CUSTOM_PRIORITY
> 	bool "Custom log thread priority"
> 	help
> 	  Set a custom value for the log processing thread.
> 
> config LOG_PROCESS_THREAD_PRIORITY
> 	int "Priority of the log processing thread"
> 	default 0
> 	depends on LOG_PROCESS_THREAD_CUSTOM_PRIORITY
> 	help
> 	  The priority of the log processing thread.
> 	  When not set the prority is set to K_LOWEST_APPLICATION_THREAD_PRIO.
> ```

Yes this is exactly what I see: without a `default` value the macro will be defined, but empty. In this case it there is no good way, that I know of, to set the default priority.
Hence I'd suggest using a binary "enable config" with the integer "priority config".

**Comment by nordic-krch** [comment]
If macro is defined but empty we could try something like
`#define THREAD_PRI COND_CODE_1(IS_EMPTY(CONFIG_LOG_CUSTOM_PRI), (K_LOWEST_APP), (CONFIG_LOG_CUSTOM_PRI))`

**Comment by justinyhuang** [comment]
> If macro is defined but empty we could try something like `#define THREAD_PRI COND_CODE_1(IS_EMPTY(CONFIG_LOG_CUSTOM_PRI), (K_LOWEST_APP), (CONFIG_LOG_CUSTOM_PRI))`

Thank you for sharing, however, I tried following the suggestion by 
` #define LOG_PROCESS_THREAD_PRIORITY COND_CODE_1(IS_EMPTY(CONFIG_LOG_PROCESS_THREAD_PRIORITY), (K_LOWEST_APPLICATION_THREAD_PRIO), (CONFIG_LOG_PROCESS_THREAD_PRIORITY))`

I still see the error when build:

` /var/folders/cb/tlt_12350v79w73_dmfks6v00000gn/T//ccO2wkIT.s: Assembler messages:
/var/folders/cb/tlt_12350v79w73_dmfks6v00000gn/T//ccO2wkIT.s:1255: Error: missing expression `

**Comment by cfriedt** [comment]
> A boolean option to depend on, might still be required?
> 
> ```
> config LOG_PROCESS_THREAD_CUSTOM_PRIORITY
> 	bool "Custom log thread priority"
> 	help
> 	  Set a custom value for the log processing thread.
> 
> config LOG_PROCESS_THREAD_PRIORITY
> 	int "Priority of the log processing thread"
> 	default 0
> 	depends on LOG_PROCESS_THREAD_CUSTOM_PRIORITY
> 	help
> 	  The priority of the log processing thread.
> 	  When not set the prority is set to K_LOWEST_APPLICATION_THREAD_PRIO.
> ```

The above sounds fine to me - I think the wording is better as well. @justinyhuang, @nordic-krch - does that work for you?

**Comment by justinyhuang** [comment]
> The above sounds fine to me - I think the wording is better as well. @justinyhuang, @nordic-krch - does that work for you?

I agree and can update the PR accordingly, if it looks good to @nordic-krch

**Review by cfriedt — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**